src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationKeyComparer.cs

```diff
@@ -29,49 +29,64 @@ public class ConfigurationKeyComparer : IComparer<string>
         /// <returns>Less than 0 if x is less than y, 0 if x is equal to y and greater than 0 if x is greater than y.</returns>
         public int Compare(string? x, string? y)
         {
-            string[] xParts = x?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-            string[] yParts = y?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-
-            // Compare each part until we get two parts that are not equal
-            for (int i = 0; i < Math.Min(xParts.Length, yParts.Length); i++)
+            if (x is not null && y is not null)
             {
-                x = xParts[i];
-                y = yParts[i];
+                if (x!.Contains(ConfigurationPath.KeyDelimiter) && y!.Contains(ConfigurationPath.KeyDelimiter))
+                {
+                    string[] xParts = x!.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries);
```

Comment:
>xParts [](http://example.com/codeflow?start=29&length=6)

Can we optimize this more but avoid the allocations occur because of string.split? I think this is possible and not hard to do it. Like use string.indexof(ordinal) then compare substrings after that.

Reply:
I agree with @tarekgh.

BTW it's a pity that the code which is consuming this type is using string to represent a complex type and uses indexof/split back and forth.

The index is known here:

https://github.com/dotnet/runtime/blob/6033ef2c56db79e68e13f2856f42d250399d927e/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationProvider.cs#L98

then a substring is performed:

https://github.com/dotnet/runtime/blob/6033ef2c56db79e68e13f2856f42d250399d927e/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationProvider.cs#L99

then all the logic is reversed to be able to sort the keys:

https://github.com/dotnet/runtime/blob/6033ef2c56db79e68e13f2856f42d250399d927e/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationProvider.cs#L91

It's just a digression, I don't expect @maryamariyan to fix it in this PR (these types are public)